Add `--format` to `docker service ls`

yongtang opened the pull request:
**- What I did**

This fix tries to improve the display of `docker service ls` and adds `--format` flag to `docker service ls`.

In addition to `--format` flag, several other improvement:
1. Adds `--no-trunc` to `docker service ls`
2. Updates `docker stacks service`.
3. Adds `servicesFormat` to config file.

Related docs has been updated.

cc @thaJeztah @vdemeester @dnephin 

**- A picture of a cute animal (not mandatory but encouraged)**

![1223](https://cloud.githubusercontent.com/assets/6932348/20130788/0f599ce0-a60e-11e6-8de3-a629939ac69d.jpg)


Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

AkihiroSuda commented:
SGTM, can you add tests?

vdemeester commented:
Design SGTM 👼

thaJeztah commented on cli/command/service/list.go:
```diff
@@ -40,6 +36,8 @@ func newListCommand(dockerCli *command.DockerCli) *cobra.Command {
 
 	flags := cmd.Flags()
 	flags.BoolVarP(&opts.quiet, "quiet", "q", false, "Only display IDs")
+	flags.BoolVar(&opts.noTrunc, "no-trunc", false, "Do not truncate the output")
```
I wonder if we should consider introducing `ShortID` and/or `TruncID` placeholders. The `TruncID` placeholder has the behavior of what `ID` has now (so, truncated by default, but show full ID if `--trunc` is added. That way people can define a template where the output _always_ has the full-id, or has an ID that can be truncated or full.

yongtang replied:
@thaJeztah does that mean we will leave `ID` alone regardless of `--no-trunc=false/true`, and only trunc `TruncID` based on `--no-trunc=true`?

thaJeztah replied:
@yongtang perhaps, yes. My initial thought was "do we need a `--no-trunc` flag? That's when I started thinking if we perhaps need other options for the format.

It's just a thought though, so perhaps @vdemeester or @dnephin has some thoughts

vdemeester replied:
Same though as @thaJeztah, I don't really like `--no-trunc` flag, especially since format uses go template. We could provide a function that shorten "or not" the ID : `--format="{{ .ID | shorten }}"` or something like that.

yongtang commented:
@AkihiroSuda Added a test for the PR.

thaJeztah commented:
Design LGTM, let's keep https://github.com/docker/docker/pull/28199#discussion_r87175650 for another discussion

vdemeester commented:
ping @yongtang, needs a rebase 👼

yongtang commented:
Thanks @vdemeester. The PR has been rebased.

AkihiroSuda commented:
```
00:00:59.032 Errors from golint:
00:00:59.032 cli/command/service/list.go:94:1: comment on exported function GetServicesStatus should be of the form "GetServicesStatus ..."
00:00:59.032 
00:00:59.032 Please fix the above errors. You can test via "golint" and commit the result.
```

vdemeester commented on cli/command/service/list.go:
```diff
@@ -107,52 +113,19 @@ func PrintNotQuiet(out io.Writer, services []swarm.Service, nodes []swarm.Node,
-			}
+			mode[service.ID] = "global"
+			replicas[service.ID] = fmt.Sprintf("%d/%d", running[service.ID], tasksNoShutdown[service.ID])
+		} else {
```
This `else` is empty right ? should go 👼

vdemeester commented on cli/command/service/list.go:
```diff
@@ -73,30 +71,38 @@ func runList(dockerCli *command.DockerCli, opts listOptions) error {
-// Besides this, command `docker stack services xxx` will call this, too.
-func PrintNotQuiet(out io.Writer, services []swarm.Service, nodes []swarm.Node, tasks []swarm.Task) {
+// GetServicesStatus returns a map of mode and replicas
+func GetServicesStatus(services []swarm.Service, nodes []swarm.Node, tasks []swarm.Task) (map[string]string, map[string]string) {
```
Should/Could this return a map of struct instead, with mode and replica in the struct, instead of two maps ? 👼

vdemeester reviewed (changes requested):
Few comments 👼 
I do agree with @thaJeztah on the `--no-trunc` flags (on other commands as well but 😅)

yongtang commented:
@vdemeester @thaJeztah Thanks for the review. The PR has been updated and now `--no-trunc` are replace with template func of `shorten`. Please take a look and let me know if there are any issues.

vdemeester reviewed (approved):
LGTM 🐸

AkihiroSuda commented on integration-cli/docker_cli_swarm_test.go:
```diff
@@ -1630,3 +1630,33 @@ func (s *DockerSwarmSuite) TestSwarmInitWithDrain(c *check.C) {
 	c.Assert(out, checker.Contains, "Drain")
 }
+
+func (s *DockerSwarmSuite) TestSwarmServiceListFormat(c *check.C) {
```
I wonder this test can be moved to UT, but it can be another PR in the future

vdemeester replied:
Yes 👍

AkihiroSuda commented on docs/reference/commandline/service_ls.md:
```diff
@@ -103,6 +104,34 @@ ID            NAME   MODE        REPLICAS  IMAGE
+`.ID`       | Service ID
+`.Name`     | Service name
+`.Mode`     | Service mode (replicated, global)
+`.Replicas` | Service replicas
```
I wonder we should have two separate placeholders e.g. `RunningReplicas` and `AllReplicas`.

WDYT?

yongtang replied:
@AkihiroSuda I feel like that means we introduce new concepts that does not necessary match the existing Table header. Should we change the table header to `RUNNING/ALL REPLICAS`? Or maybe we could leave it alone for now?

AkihiroSuda replied:
I think it is ok to introduce all `.Replicas`, `.RunningReplicas`, and `.AllReplicas`.
But it is ok to leave it alone for now.

yongtang commented:
@AkihiroSuda @vdemeester Thanks for the review. The PR has been updated with integration tests have been replaced by unit tests. Please take a look.

AkihiroSuda reviewed (approved):
LGTM

PTAL @vdemeester @thaJeztah for confirmation.

thaJeztah commented:
Let me put it in code-review for a bit, I'd like to confer with other maintainers on the `shorten` approach from a UX consistency perspective. I think it's useful, but need to be sure before we merge :heart:

yongtang commented:
@thaJeztah in case we want to stick with `shorten` we may want to consider deprecated `--no-trunc` for other commands (for consistency). Either is fine for me though.

thaJeztah commented:
@yongtang hm, okay discussed this in the maintainers meeting, and we think it would be better to put the `shorten` option in a separate PR, and make this PR just add the regular "formatting" option, so we can discuss the `shorten` option a bit more in depth

(sorry :smile:)

yongtang committed:
Add `--format` to `docker service ls`
This fix tries to improve the display of `docker service ls`
and adds `--format` flag to `docker service ls`.

In addition to `--format` flag, several other improvement:
1. Updates `docker stacks service`.
2. Adds `servicesFormat` to config file.

Related docs has been updated.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @thaJeztah. The PR has been updated. Also created another PR #30484 for `shorten`. Please take a look.

Also in this PR I didn't add the `--no-trunc` flag and the output of ID will be truncated by default (same as old behavior). Please let me know if we want to introduce `--no-trunc` in this PR, or we want to have the full ID output by default (behavior will be different from original).

vdemeester reviewed (approved):
LGTM 🐯

vdemeester commented:
Moving to docs-review
/cc @thaJeztah

vdemeester reviewed (changes requested):
@yongtang  Could you update `docs/reference/commandline/cli.md` to add an example for `ServicesFormat` ?

yongtang commented:
Thanks @vdemeester. The docs has been updated in the PR. Please take a look and let me know if there are any issues.

vdemeester reviewed (approved):
docs LGTM 🐸
/cc @thaJeztah for docs revisit
